src/core/ad.jl

```diff
@@ -95,28 +99,32 @@ using forwards-mode AD from ForwardDiff.jl.
 """
 function gradient_forward(
     θ::AbstractVector{<:Real},
-    vi::VarInfo,
+    vi::AbstractVarInfo,
     model::Model,
     sampler::Union{Nothing, Sampler}=nothing,
 )
     # Record old parameters.
-    vals_old, logp_old = copy(vi.vals), copy(vi.logp)
+    vals_old, logp_old = copy(vi[sampler]), copy(vi.logp)
 
     # Define function to compute log joint.
     function f(θ)
-        vi[sampler] = θ
-        return -runmodel!(model, vi, sampler).logp
+        new_vi = NewVarInfo(vi, sampler, eltype(θ))
```

Comment:
Does `NewVarInfo` reallocate memory here or it just re-type things?

Reply:
It re-allocates. This is a performance issue which can be improved. I tried memoizing the output of `NewVarInfo` using `Memoize.jl` but it was slower because it introduced a type instability, and allocating is pretty fast. Perhaps the best option here would be to define the function `f` as a functor, and try to infer the AD type in advance, then allocate the AD vector as part of `vi`. This is a different surgical operation though so let's not do heart and brain surgery at the same time. I would be happy if we can just get this through without any repercussions elsewhere in the package, since I am touching quite a bit of Turing here. I am yet to add proper tests for `TypedVarInfo` to match those of `UntypedVarInfo` and compare the 2 together.